Lightweight Storefront: Update theme setting screen to display theme list

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #11337 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR integrates the API to fetch suggested themes and adds the themes carousel to the theme setting screens. Changes include:
- `ThemeSettingView`: Added theme carousel view in place.
- `ThemeSettingViewModel`: added a reference to the view model of the carousel view to load themes after getting the current theme.
- Added a view model for `ThemesCarouselView` and integrated the themes API for loading real data.
- Updated the `ThemesCarouselView` with different states based on the load request.
- Updated the Settings screen to display items in the correct order following the design.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
- Log in to a WPCom store and navigate to the Menu tab.
- Select Settings. If the store hasn't completed onboarding tasks, notice that the Themes item is below the option to hide the onboarding task list.
- Select Themes. Confirm that the screen displays the suggested themes as filtered in `WordPressThemeRemote`. If your store has one of the suggested theme, the theme should not be displayed on the list.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->



https://github.com/woocommerce/woocommerce-ios/assets/5533851/a784460a-b706-4934-a368-c9262e3f0099




---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

itsmeichigo committed:
Add theme carousel view to theme setting view

itsmeichigo committed:
Add view model for ThemeCarouselView

itsmeichigo committed:
Inject carousel view model to views

itsmeichigo committed:
Integrate loading suggested themes

itsmeichigo committed:
Update carousel view with real data

itsmeichigo committed:
Add error state

itsmeichigo committed:
Update loading and error state views sizes

itsmeichigo committed:
Fix issue loading theme image

itsmeichigo committed:
Update settings order

itsmeichigo committed:
Log error loading themes

itsmeichigo committed:
Rename ThemesCarouselViewModel and add test file

itsmeichigo committed:
Update tests for ThemesCarouselViewModel

itsmeichigo committed:
Filter current theme from list

itsmeichigo committed:
Silence hound

itsmeichigo committed:
Add test for filtering matching theme id

itsmeichigo committed:
Update ThemesCarouselViewModelTests.swift

itsmeichigo committed:
Clean up code for updateCurrentThemeName

itsmeichigo committed:
Update tests for ThemeSettingViewModel

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Themes/ThemesCarouselView.swift:
```diff
@@ -1,52 +1,89 @@
 
     /// The content of the message at the end of the carousel.
-    let lastMessageContent: String
+    private let lastMessageContent: String
```
Opinion: 

Can we move these strings to view model? 

Additionally, we could even add a `Mode` enum with `settings` and `storeCreation` to the view model to change these strings based on the mode. 

This way, we can inject the mode and let the view model pick the strings.

itsmeichigo committed:
Fix failed unit test

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Themes/ThemesCarouselView.swift:
```diff
@@ -1,52 +1,89 @@
+                .secondaryBodyStyle()
+            Button {
+                Task {
+                    await viewModel.fetchThemes()
```
Are we losing the `currentThemeID` information when retrying?

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Themes/ThemesCarouselViewModel.swift:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @MainActor
+    func fetchThemes(currentThemeID: String? = nil) async {
```
❓  Is it possible to move the logic of `currentThemeID` out of this method? 

This way we don't have to wait for the load current theme network request. i.e. We can load the current theme and load suggested themes in parallel. 

What do you think?

itsmeichigo replied:
I did this at first, but the theme list result arrives faster, so it'll be filtered after the current theme is loaded. This causes the filtered theme to appear briefly then disappear - which doesn't look nice. We can cache the current theme ID to UserDefaults so that the weird animation happens only on the first request, but it complicates the logic a bit more. WDYT @selanthiraiyan?

selanthiraiyan replied:
Thanks for the explanation, Huong! 

I guess we have to separate the fetch and render process into separate methods to achieve a smooth experience. The extra complexity isn't worth it for the themes screen, considering we don't expect the user to access this theme setting often.

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Themes/ThemesCarouselViewModelTests.swift:
```diff
@@ -0,0 +1,82 @@
+@testable import Yosemite
+
+@MainActor
+final class ThemesCarouselViewModelTests: XCTestCase {
```
Nice tests. ❤️

selanthiraiyan reviewed (approved):
This works as expected. 🎉 

I left a few non-blocking questions and comments for your consideration.

itsmeichigo committed:
Update logic for fetching themes

itsmeichigo committed:
Load themes on the carousel view

itsmeichigo committed:
Update tests for loading themes

itsmeichigo committed:
Inject new param mode to ThemesCarouselViewModel

itsmeichigo committed:
Move last message card texts to theme carousel mode

itsmeichigo committed:
Update state immediately if the carousel mode is profiler

itsmeichigo committed:
Update tests for ThemeCarouselViewModel

itsmeichigo commented:
Thank you @selanthiraiyan for the reviews and suggestions! I added changes with a new parameter `mode` for the `ThemesCarouselViewModel` and updated the logic for loading themes. Please take another look when you can 🙏

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Themes/ThemesCarouselViewModel.swift:
```diff
@@ -0,0 +1,111 @@
+    init(mode: Mode, stores: StoresManager = ServiceLocator.stores) {
+        self.mode = mode
+        self.stores = stores
+        observeThemeListAndCurrentTheme()
```
❓ Would it make sense to observe only if `mode` is `themeSettings`? This is to avoid setting `state` from two places for `storeCreationProfiler` mode. Please correct me if I am wrong.

itsmeichigo replied:
Good idea! Updated in a043a7a0a39868c00ca5b2092d108bd1655027e5.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Themes/ThemesCarouselViewModel.swift:
```diff
@@ -0,0 +1,111 @@
+}
+
+private extension ThemesCarouselViewModel {
+    func observeThemeListAndCurrentTheme() {
```
Opinion: What do you think about renaming it to reflect the purpose of this method? I have `waitForCurrentThemeAndFinishLoading` in mind.

itsmeichigo replied:
Also handled this in a043a7a0a39868c00ca5b2092d108bd1655027e5.

selanthiraiyan reviewed (approved):
Thank you for the updates, Huong!

Nice solution. 👍

itsmeichigo committed:
Rename observeThemeListAndCurrentTheme and update conditional check

itsmeichigo committed:
Merge branch 'trunk' into issue/11337-theme-picker-settings

itsmeichigo commented:
Thanks Sharma, I've added some last changes following your suggestions. I'm merging the PR to proceed with the next task.

itsmeichigo committed:
Fix failed test
